deepspeed/module_inject/replace_module.py

```diff
@@ -562,6 +564,14 @@ def replace_module(model, orig_class, replace_fn, _replace_policy, checkpoint=No
         if embedding_weight is not None and hasattr(replaced_module, "lm_head") and hasattr(
                 replaced_module.lm_head, "weight") and replaced_module.lm_head.weight.is_meta:
             replaced_module.lm_head.weight = embedding_weight
+
+    # enable tensor parallel for the last linear
+    if hasattr(replaced_module, "lm_head") and hasattr(replaced_module.lm_head,
+                                                       "weight") and not replaced_module.lm_head.weight.is_meta:
+        replaced_module = replace_fn(replaced_module, ("lm_head", ), 0, "lm_head")
+    elif hasattr(replaced_module, "embed_out") and hasattr(replaced_module.embed_out,
+                                                           "weight") and not replaced_module.embed_out.weight.is_meta:
+        replaced_module = replace_fn(replaced_module, ("embed_out", ), 0, "embed_out")
```

Comment:
Hi @RezaYazdaniAminabadi @delock, I refactored some code here, so that it will not integrate so deeply with former replace logic. And it will replace the `lm_head` or `embed_out` Linear **after** all other replace/load logics are finished. It will be more decoupling and cleaner now.